opendevin/core/Message.py

```diff
@@ -0,0 +1,29 @@
+from typing import List, Optional
+
+from pydantic import BaseModel, Field, model_serializer
+from typing_extensions import Literal
+
+
+class Message(BaseModel):
+    role: Literal['user', 'system']
+    image_urls: Optional[List[str]] = Field(default=None)
+    text: str
```

Comment:
I wonder how would this handles the case where we have image-text interleave data?

For example, an user *might* send:
```
This is what it used to look like:
<img1>
After i apply YYY, it becomes:
<img2>
```
I get that our current front-end design doesn't really support this interleaved data format, but i'd hope we could make the backend implementation more general -- in case we want to support this interleave feature in the future.


Maybe we could follow the standard of API call:
https://litellm.vercel.app/docs/completion/vision#quick-start

```
import os 
from litellm import completion

os.environ["OPENAI_API_KEY"] = "your-api-key"

# openai call
response = completion(
    model = "gpt-4-vision-preview", 
    messages=[
        {
            "role": "user",
            "content": [
                            {
                                "type": "text",
                                "text": "What’s in this image?"
                            },
                            {
                                "type": "image_url",
                                "image_url": {
                                "url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg"
                                }
                            }
                        ]
        }
    ],
)
```


So that:
```
class ContentType(Enum):
    TEXT = 'text'
    IMAGE_URL = 'image_url'

class TextContent(BaseModel):
    type = ContentType.TEXT
    content: str

class VisionContent(BaseModel):
    type = ContentType.IMAGE_URL
    image_urls: dict[str, str]

class Message(BaseModel):
    role: Literal['user', 'system']
    content: list[TextContent|ImageContent] = Field(default=list)
```

Reply:
the thing here is that, i feel only for one text  is read by the model(the litellm api), so maybe this has to be a frontend change or something where we send them as two messages. .

Reply:
i think here the image_urls has to also be a list, so that we can do:  text, img, text img img

Reply:
i am thinking of doing this:
```

class ContentType(Enum):
    TEXT = 'text'
    IMAGE_URL = 'image_url'

class TextContent(BaseModel):
    type: ContentType = ContentType.TEXT
    text: str

    @model_serializer
    def serialize_model(self):
        return { 'type': self.type.value, 'text': self.text}

class ImageContent(BaseModel):
    type: ContentType = ContentType.IMAGE_URL
    image_urls: list[str]

    @model_serializer
    def serialize_model(self):
        images: list[dict[str, dict[str, str]]] = []
        for url in self.image_urls:
            images.append({'type': self.type.value, 'image_url': {'url': url}})
        return images

class Message(BaseModel):
    role: Literal['user', 'system']
    content: list[TextContent|ImageContent] = Field(default=list)

    @property
    def contains_image(self) -> bool:
        return bool(self.image_urls and len(self.image_urls) > 0)

    @model_serializer
    def serialize_model(self) -> dict:
        content: list[dict[str, str | dict[str, str]]] = []

        for item in self.content:
            if isinstance(item, TextContent):
                content.append(item.model_dump())
            elif isinstance(item, ImageContent):
                content.extend(item.model_dump())
        
        return content

text_content1 = TextContent(text="This is a text message")
image_content1 = ImageContent(image_urls=["http://example.com/image1.png", "http://example.com/image2.png"])

text_content2 = TextContent(text="This is a text message")
image_content2 = ImageContent(image_urls=["http://example.com/image3.png", "http://example.com/image4.png"])

message = Message(role='user', content=[text_content1, image_content1, text_content2, image_content2])
print(message.serialize_model())
```

Reply:
Maybe it would be great if someone would write the tests for this once we are done.

Reply:
how do you want to handle this: 
```
def get_current_user_intent(self):
        """Returns the latest user message and image(if provided) that appears after a FinishAction, or the first (the task) if nothing was finished yet."""
        last_user_message = None
        for event in self.history.get_events(reverse=True):
            if isinstance(event, MessageAction) and event.source == 'user':
                last_user_message = Message(
                    role='user', text=event.content, image_urls=event.images_base64
                )
            elif isinstance(event, AgentFinishAction):
                if last_user_message is not None:
                    return last_user_message

        return last_user_message
```
```
    def step(self, state: State) -> Action:
        last_user_message = state.get_current_user_intent()
        prompt = self.prompt_template.render(
            state=state,
            instructions=instructions,
            to_json=to_json,
            history_to_json=self.history_to_json,
            delegates=self.delegates,
            latest_user_message=last_user_message.text if last_user_message else None,
        )
        
        message = Message(
            role='user', text=prompt, image_urls=last_user_message.image_urls if last_user_message else None
        )
        resp = self.llm.completion(messages=[message.model_dump()])
        action_resp = resp['choices'][0]['message']['content']
        action = parse_response(action_resp)
        return action
```

I was returning a Message object but i am changing it so bascially i am just using it to return data and the Message object in the get_current_user_intent is not of much use.  I don't understand what the get_current_user_intent does.

Reply:
hmm, could you return the "Message" directly to be more extensible for future use? Or i guess it doesn't really matter that much for the context of this PR - feel free to do whatever you think is the easiest / cleanest

Reply:
@Kaushikdkrikhanu This method attempts to figure out what is the user message which best expresses the current task. That is, in opendevin in an interactive session, we can have:
- user message saying "create this code etc"
- agent doing stuff
- sometimes user message saying something
- agent doing stuff or saying stuff

This method tries to figure out what's the task. It would, in this scenario, return the first message, not the rest of user messages.

Later, a sequence might look like this:
- user message
- agent doing stuff or saying stuff
- agent sending AgentFinishAction - it does that when it thinks it solved the current task
- user message (e.g. "ok, but can you refactor this that way?")
- agent ... etc

In this scenario, the most likely current user intent will be detected as the first user message after a Finish action, so the refactor stuff.

Obviously, this is far from perfect. It works well in simple cases and not so well when users chat freely 😅. Nevertheless, it's pretty important in the execution flow. IMHO we should include in it the images.